class-two-factor-core.php

```diff
@@ -673,21 +677,30 @@ public static function destroy_current_session_for_user( $user ) {
 	}
 
 	/**
-	 * Prevent login through XML-RPC and REST API for users with at least one
-	 * two-factor method enabled.
+	 * Trigget the two-factor workflow only for valid login attempts
+	 * with username present. Prevent authentication during API requests
+	 * unless explicitly enabled for the user (disabled by default).
 	 *
-	 * @param  WP_User|WP_Error $user Valid WP_User only if the previous filters
+	 * @param WP_User|WP_Error $user Valid WP_User only if the previous filters
 	 *                                have verified and confirmed the
 	 *                                authentication credentials.
+	 * @param string           $username The username.
+	 * @param string           $password The password.
 	 *
 	 * @return WP_User|WP_Error
 	 */
-	public static function filter_authenticate( $user ) {
-		if ( $user instanceof WP_User && self::is_api_request() && self::is_user_using_two_factor( $user->ID ) && ! self::is_user_api_login_enabled( $user->ID ) ) {
-			return new WP_Error(
-				'invalid_application_credentials',
-				__( 'Error: API login for user disabled.', 'two-factor' )
-			);
+	public static function filter_authenticate( $user, $username, $password ) {
+		if ( strlen( $username ) && $user instanceof WP_User && self::is_user_using_two_factor( $user->ID ) ) {
```

Comment:
The main difference here is that we're now triggering any of this logic only for requests with the `$username` supplied which indicates a login attempt instead of a simple authenticated request to `wp-login.php`.